### PR TITLE
[Sol->Yul] Adding util function to copy literal to storage.

### LINKED ISSUE #10349: Sol->Yul: string literals should not be converted to memory before stored in storage
Assigning a string literal to a storage value should not need conversion to memory before.

There are several similar situations:

 - abi.encode("abc")
 - contract c { bytes m = "abc"; }
 - contract c { bytes m; function f() public pure { m = "abc"; }
 - x.push("abc")
 - in https://github.com/ethereum/solidity/pull/10759 we copy a literal to memory first before copying it to storage again.

### PR DESCRIPTION
**Pull request by mijovic**
Depends on #11271.
Fixes #10349
Only util function is implemented and should be used everywhere. Right now it is used only in one place...

### CODE REVIEW

File: libsolidity/codegen/YulUtilFunctions.h
Comment (mijovic) [inline libsolidity/codegen/YulUtilFunctions.h]:
```suggestion
	/// @returns the name of a function that stores a string literal at a specific location in storage
```

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Will this be called in the generic "copy to storage" function?

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Do we have to truncate dst first?

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `029250a238e877aecf9da48cf7748d7e8a8e9db4`:
```
Coding style error:
 libsolidity/codegen/YulUtilFunctions.cpp:155:std::string YulUtilFunctions::copyLiteralToStorageFunction(const string &_literal)
 libsolidity/codegen/YulUtilFunctions.cpp:155:std::string YulUtilFunctions::copyLiteralToStorageFunction(const string &_literal)

```
Please check that your changes are working as intended.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (stackenbotten) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Coding style error

Comment (stackenbotten) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Coding style error

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Done

**Comment by mijovic** [comment]
Conversions to memory happen as the `closestTemporaryType` for `StringLiteralType` in case when we assign it to storage is `bytes memory`
So we need to do copy to memory and then copy to storage for both:
- contract c { bytes m = "abc"; }
- contract c { bytes m; function f() public pure { m = "abc"; }

I can remove unnecessary conversion by implementing specfic `closestTemporaryType` for `StringLiteralType` and then we will end up in `copyToStorage` generic function and we will be able to copy literal directly to storage.
Problem with this is that it will break old codegen which I am fine with if I just call IR implementation to copy literal to storage.

Any other opinion or suggestion @ekpyron @chriseth ?

**Comment by chriseth** [comment]
Doesn't the old codegen avoid conversion to closest temporary type?

**Comment by chriseth** [comment]
I had a quick look through the code, and I have the impression that the whole mobileType / closestTemporaryType mechanism can be reworked. It might have served a better purpose in the past, but it seems it's mostly used for RationalNumberTypes.

Maybe we can also get rid of the 'isPointer' property of reference types - the lvalue mechanism should actually already cover this.

It might be possible to just remove the conversion in the assignment, and I think we should also implement a new function similar to mobileType / closestTemporaryType and remove the other two.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Can you write this as `if(isShiftOp) solAssert(...)`?

**Comment by chriseth** [comment]
On a superficial review, this looks good!

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (mijovic) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Will do like that

**Comment by mijovic** [comment]
> On a superficial review, this looks good!

I need to figure out why compound operators are broken with this change in assignment operator. Than it will be ready.
Also, good thing is that this PR does also optimize copy of `calldata slices` to `storage` in `IR`, as before it was converting slices first to `memory`, than from `memory` to `storage` as the `closestTemporaryType` for slice was `memory array`

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (ekpyron) [inline libsolidity/codegen/YulUtilFunctions.cpp]: For cases like this I'm wondering *every* time, if it wasn't easier to read to split them up. Here for example split into a `long` and a `short` case with two templates... I'm not saying we should necessarily do that here, but I would be curious about the general opinion: is using one template with more case variables really preferable to two templates (maybe as `<body>` in the outer function template)?
I personally think I would have review a split version twice as fast...

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (ekpyron) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Isn't the length already written to in the resize function? Maybe we could even go as far as to *not* calling the resize function, but just manually clearing the tail with ``clearStorageRange`` when decreasing the size here... but it's not nice... not quite sure, but I'm wondering.

Maybe using the resize function and just writing the length again in the "short" case is fine, but the write in the long case we can definitely skip, can't we?

**Review by ekpyron — COMMENTED** [review]
Some comments.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Hm, indeed, you are right. It is not easiest to review as is! Also, it looks much simpler if I split into different cases

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: We can skip writing length in case of long one, and I did it like that for now

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (ekpyron) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Yeah, the split version is much nicer indeed now. I wonder whether we should have some helper for doing this boilerplate here, though. Maybe even a toplevel member of ``m_functionCollector``, s.t. this would read
```
m_functionCollector.createFunctionWithArgsAndBody(functionName, [&](auto& args, auto& body) {
  args = "slot";
  body = Whiskers(...)(...).render();
});
```
(Or with another similar mechanism, the named output arguments were just a quick and dirty idea right now.)
But: different PR for sure.

Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: I think that makes a lot of sense, as many (if not all of calls to `createFunction`) start with:
```
string functionName = "...";
return m_functionCollector.createFunction(functionName, [&]() {
	Whiskers templ(R"(
		function <functionName>(<args>) -> <returns> {
```

It would simplify not only places where we have `<body>` template :)

Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: I'll open issue for this now.
Edit: #11049

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (ekpyron) [inline libsolidity/codegen/YulUtilFunctions.cpp]:
```suggestion
					let dstPtr := <dataArea>(slot)
```
?

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (ekpyron) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Ah, right, actually since `ẁordValue` and `length` are both constant this should reduce to a simple write after optimization plus the optimizer can maybe even remove the write to the length in the resize above, so this is good!

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (ekpyron) [inline libsolidity/codegen/YulUtilFunctions.cpp:2746]: What's this about array slices? Does this belong here or did this sneak in by accident?
I mean, the change looks good anyways, just doesn't seem that related to the rest...

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (ekpyron) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:413]: For the record, I still need to verify the types and conversions here :). So nothing `closestTemporaryType` does is required here and above?

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp:2746]: This was interesting, and I am glad you asked.
closestTemporaryType for calldata slices (when converting to storage array) was memory array and there was "hidden" conversion from calldata to memory and then from memory to storage.

Comment (ekpyron) [reply libsolidity/codegen/YulUtilFunctions.cpp:2746]: Interesting indeed! Not sure how that's happening... ``Type::closestTemporaryType`` seems to return ``mobileType``, if `_targetType` is in storage... and ``ArraySliceType::mobileType`` looks to me as if it returns the underlying array type which should be in calldata... so not sure right now what's going on there :-).

Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp:2746]: Indeed, you are right. It was calldata array comming here before, and now, since `closestTemporaryType` is not used anymore here we get actual slice instead. That is why these changes were needed....

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (mijovic) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:413]: I spent a lot of time to check these conversions, they look good to me.
So, nothing is required from `closestTemporaryType` both here and above.

But, doesn't hurt to get it double checked :)

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: I did add a quick and first draft version of yul function collector update. It is here #11066.
Once that, or something similar gets merged, we could use that interface to update this one

**Comment by mijovic** [comment]
Rebased, and inlined usage of `closestTemporaryType` in legacy codegen.

**Comment by ekpyron** [comment]
The test failures are gas cost changes - surprising ones actually, seems like this can tip both ways.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: It would be interesting to see how this is optimized.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp:2774]: Why don't we pass the type itself?

Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp:2774]: If it stays like this, can you add an assertion and comment that this only works properly for calldata array slices?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: If this includes an implicit conversion, don't we do the conversion twice?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Can you use `else` instead? These two lines are duplicated.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Can you use `writeToMemoryfunction`?

File: test/libsolidity/semanticTests/array/copying/copy_byte_array_to_storage.sol
Comment (chriseth) [inline test/libsolidity/semanticTests/array/copying/copy_byte_array_to_storage.sol]: Why does it get so much more expensive? Shouldn't it get cheaper?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (mijovic) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Having multiple conversions maybe explains gas cost increase. Will investigate this.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: So this ends up being (on default settings for optimizer):
```
function copy_literal_to_storage_4e03657aea45a94fc7d47ba826c8d667c0d1e6e33a64a036ec44f58fa12d6c45(slot)
{
    resize_array_t_bytes_storage_ptr(slot, 3)
    sstore(slot, add("abc", mul(2, 3)))
}
```
Also, this resize function has two cases where it calls different function for increasing and decreasing size where the actual write of length is happening. Optimizing `sstore`s depends on if function would be inlined or not. I think it was mentioned in #10479 

What are the chances for optimizer to optimize this out soon? @hrkrshnn @chriseth @ekpyron 
If it is unlikely, maybe better doing some inlined checks here and calling some kind of `clearStorageRange` function instead of full resize call.

`mul(2, 3)` is obviously compile time constant and I have changed it now so we set it to 6 in compile time...

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Before #10479 we will be able to optimize out some `sstore`s.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (mijovic) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Fixed

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (mijovic) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Done

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp:2774]: Added

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: This performs a type conversion.

**Comment by chriseth** [comment]
Please be careful, you are not doing a mere refactoring of the IR codegen here. Before, we first converted the right hand side and then visited the left hand side.

**Comment by mijovic** [comment]
> Please be careful, you are not doing a mere refactoring of the IR codegen here. Before, we first converted the right hand side and then visited the left hand side.

I moved conversion upper, where it was before, just before visiting left hand side.
For reference types I didn't do conversion, which leads to direct conversion to target type which should be fine?

File: test/libsolidity/semanticTests/array/copying/copy_byte_array_to_storage.sol
Comment (chriseth) [reply test/libsolidity/semanticTests/array/copying/copy_byte_array_to_storage.sol]: I had a look at the diff but could not really make out where the change comes from.

**Review by chriseth — DISMISSED** [review]

**Comment by mijovic** [comment]
I just did a rebase on top of latest develop

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (mijovic) [inline libsolidity/codegen/YulUtilFunctions.cpp]: There is an `sstore` inside of resizeArray function, and combined with `sstore` just below, we have multiple `sstore`s to the same slot in case of short literal. Before, we copied literal to memory and then we did copy to storage. In `copyByteArrayToStorage` function, we do not call directly `resizeArray` but we do truncation there  which doesn't produce two `sstore`s. This leads to having more gas used when we assign literal to storage array.

And in `copy_byte_array_to_storage` test, we have multiple similar assignments `data = "abc";`

File: test/libsolidity/semanticTests/array/copying/copy_byte_array_to_storage.sol
Comment (mijovic) [reply test/libsolidity/semanticTests/array/copying/copy_byte_array_to_storage.sol]: I think this is the reason https://github.com/ethereum/solidity/pull/10768/files#r610482903

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Oh, even more `sstore`s are waisted here.
I'll try explaining what else happens. So, we have for example part of the test mentioned:
```
data = "abc";
if (!(readDataSlot() == zero)) return 2;
data = "1234567890123456789012345678901234567890123456789012345678901234567890";
```
data is resized from short to long byte array. So, `resizeArray` copies "abc" to new data location... That data is being overwritten as the array is just being overwritten here. I would suggest adding flag to `resizeArray` to distinguish type of resize we want to have.

It looks like this type of resize is working only for push. 

Nothing of this is done in copying byte array from memory to storage...

**Comment by mijovic** [comment]
Last commit I have added introduced resize mode for arrays. In case of rewriting array which happens when we do assignment, it decreases gas cost. For now It is used only in case of copying literal to storage, and shows actual gas improvements...
If that is a good direction I can finish resizeArray function properly.

**Comment by chriseth** [comment]
Splitting up the array size change functions is a very good idea, but I think using a flag makes this rather difficult to understand.

Maybe we should have a fully independent `delete array` function and could the push/pop versions be implemented by calling `increment/decrement` directly?

**Comment by mijovic** [comment]
> Splitting up the array size change functions is a very good idea, but I think using a flag makes this rather difficult to understand.
> 
> Maybe we should have a fully independent `delete array` function and could the push/pop versions be implemented by calling `increment/decrement` directly?

Makes sense. I will create `deleteArray` function that will delete elements of array up to some length. In case of delete it would be `0`, and in case of overwrite it will be `new length`. In case of push pop we can call special increment decrement functions.

**Comment by chriseth** [comment]
Should delete (I mean delete everything) be its own function?

**Comment by mijovic** [comment]
Converting to draft until #11266 is merged

**Comment by axic** [comment]
@mijovic this PR still points to the branch of #11266, which has been closed. What is the fate of this PR?

**Comment by mijovic** [comment]
> @mijovic this PR still points to the branch of #11266, which has been closed. What is the fate of this PR?

It is waiting for this one https://github.com/ethereum/solidity/pull/11271

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (ekpyron) [inline libsolidity/codegen/ExpressionCompiler.cpp]:
```suggestion
	auto closestType = [op, binOp, &closestType](
```

The type of a lambda is *not* ``std::function`` and ``std::function`` has serious overhead compared to the lambda-closure type itself.

**Review by ekpyron — DISMISSED** [review]

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (mijovic) [reply libsolidity/codegen/ExpressionCompiler.cpp]: I created file level helper. It doesn't work with `auto` as it is recursive and needs type in order to be passed via capture list...

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Can you rearrange the cases so that there is no negation? `else if (auto tupleType = dynamic_cast<...>)` could also be shorter.

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp:64]: This could actually use `applyMap`, couldn't it?

Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp:64]: Ah sorry, it cannot.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]:
```suggestion
			("copyToStorage", copyLiteralToStorageFunction(dynamic_cast<StringLiteralType const&>(_fromType).value()))
```

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (mijovic) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Done

**Review by chriseth — APPROVED** [review]

**Comment by chriseth** [comment]
@ekpyron this is not merging because it is waiting for approval from you

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (ekpyron) [reply libsolidity/codegen/ExpressionCompiler.cpp]: There's a funny trick to define local recursive lambdas: you don't have them capture themselves, but you pass them to themselves as arguments :-)
```
auto fib = [](int _n, auto&& _recurse) { return _n < 2 ? _n : _recurse(_n - 1) + _recurse(_n - 2); }
int n = fib(1, fib);
```
or even:
```
auto fib = [](int _n) {
  auto fibImpl = [](int _n, auto&& _recurse) { return _n < 2 ? _n : _recurse(_n - 1) + _recurse(_n - 2); }
  return fibImpl(_n, fibImpl);
};
int n = fib(1);
```

But file-level sounds better here anyways :-).

Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp]: `s/funny/horrible/`

Comment (ekpyron) [reply libsolidity/codegen/ExpressionCompiler.cpp]: I don't know what's not to like about it :-).

Comment (ekpyron) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Especially compared to what that innocent looking constructor of an ``std::function`` actually does, if one considers that as an alternative :-). That involves creating type-erased management objects with a v-table that has virtual destructor and copy functions and what-not... whereas the recursive lambda should actually result in the same code as a proper function.
But sure: most of the time a proper function is probably generally nicer :-).

**Pull request merged**
